backend/client/text_captions.py

```diff
@@ -1,64 +1,114 @@
 from backend.client import Client
-from backend.tasks import Task, TaskType
+from backend.tasks import Task
 from backend.client.states import ClientState as CS
+from backend.data.text_captions import TextCaptionBatchElement
+
+import gradio as gr
 
 class TextCaptionClient(Client):
     def __init__(self, id : int):
-        super(Client, self).__init__(self.id)
-
-        self.waiting = False # Waiting on frontend
-
-        self.json_buffer = None
-
-    def handle_task(self) -> bool:
-
-        # Can only handle a task if we have one
-        assert self.task is not None
-
-        if not self.waiting:
-            # Create JSON payload from the information in the task that will be presentable to the user
+        super().__init__(id)
 
-            data = {}
+        self.front = None
 
-            data['story'] = self.task.data.text
-            data['caption_index'] = self.task.data.caption_index
-            data['captions'] = self.task.data.captions
-
-            self.send_json(data)
-            self.waiting = True
-        
-            return False
-        elif self.json_buffer is not None:
-            # user is done and wants to send caption back
-            data = self.json_buffer
-            self.json_buffer = None
-
-            self.task.data.story = data['story']
-            self.task.data.caption_index = data['caption_index']
-            self.task.data.captions = data['captions']
-
-            self.task.receiver = TaskType.PIPELINE
-            self.task.sender = TaskType.USER
-
-            # We have everything we need, no longer busy or waiting
-            # Ready to send finished task back to orchestrator
-            self.waiting = False
-            self.json_buffer = None
-            self.state = CS.IDLE
-
-            return True
-        else:
-            return False
-
-    def send_json(self, data):
-        # TODO
-        pass
-
-    def receive_json(self, data):
-        # TODO 
-        self.json_buffer = data
+    def push_task(self, task : Task):
+        try:
+            assert self.front is not None
+        except:
+            raise Exception("Error: Pushed task to frontend before it was initialized")
         
+        self.task = task
+        data : TextCaptionBatchElement = task.data
+        self.front.update(data)
+
+    def init_front(self) -> str:
+        self.front = TextCaptionFront()
+        self.front.client = self
+
+        return self.front.url
+
+    def front_ping(self):
+        """
+        For frontend to ping client that it is done with task.
+        """
+        # First update task data
+        self.task.data = self.front.data
+        self.notify() 
         
+class TextCaptionFront:

```

Comment:
Can we really not abstract away any of this boiler plate?

Reply:
I need to learn more about Gradio tbh. You're right that it is extremely boiler plate cause I was changing it a lot as I went and am still not sure what features we'd need in every front end variation. Once I have a better sense of that, I can abstract front ends in general and move a lot of the specific functionality in TextCaptionClient to Client.

Reply:
Do you want to merge this now and then abstract in the future?

Reply:
Yep! That'll probably actually be the next thing I do.